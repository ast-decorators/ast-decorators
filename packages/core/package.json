{
  "name": "@ast-decorators/core",
  "version": "0.1.0",
  "description": "AST Decorators core plugin",
  "main": "lib/index.js",
  "scripts": {
    "build": "npm run build:remove && run-p build:js build:dts",
    "build:js": "babel src -d lib -x .ts --config-file ../../babel.config.json",
    "build:dts": "tsc -p tsconfig.build.json --emitDeclarationOnly",
    "build:remove": "rimraf lib",
    "typecheck": "tsc -p tsconfig.json --noEmit"
  },
  "dependencies": {
    "@babel/types": "^7"
  }
}
