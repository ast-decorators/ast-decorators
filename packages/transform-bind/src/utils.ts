import ASTDecoratorsError from '@ast-decorators/utils/lib/ASTDecoratorsError';
import {ClassMember} from '@ast-decorators/utils/lib/common';
import {isClassMethod, isClassPrivateMethod} from '@babel/types';

export type TransformBindOptions = Readonly<{
  transformerPath?: string;
}>;

export const assert = (decorator: string, member: ClassMember): void => {
  if (!isClassMethod(member) && !isClassPrivateMethod(member)) {
    throw new ASTDecoratorsError(
      `Applying @${decorator} decorator to something other than method is not allowed`,
    );
  }
};
