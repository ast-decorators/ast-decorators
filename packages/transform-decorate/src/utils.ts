export type TransformBindOptions = Readonly<{
  transformerPath?: string;
}>;
