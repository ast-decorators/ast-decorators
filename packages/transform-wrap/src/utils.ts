import ASTDecoratorsError from '@ast-decorators/utils/lib/ASTDecoratorsError';
import getMemberName from '@ast-decorators/utils/lib/getMemberName';
import type {ClassMemberMethod} from '@ast-decorators/utils/src/common';
import {
  ArrowFunctionExpression,
  FunctionExpression,
  Identifier,
  isClassPrivateProperty,
  isClassProperty,
  isLiteral,
  MemberExpression,
} from '@babel/types';

export type TransformWrapOptions = Readonly<{
  transformerPath?: string;
}>;

export type AllowedWrappers =
  | FunctionExpression
  | ArrowFunctionExpression
  | Identifier
  | MemberExpression;

export const assertWrap = (
  wrapperFunction: AllowedWrappers | undefined,
  member: ClassMemberMethod,
): void => {
  if (!wrapperFunction) {
    const methodName = getMemberName(member);

    throw new ASTDecoratorsError(
      `@wrap: No wrapper function provided${
        methodName !== undefined ? ` for ${methodName}` : ''
      }`,
    );
  }

  if (
    (isClassProperty(member) || isClassPrivateProperty(member)) &&
    (!member.value || isLiteral(member.value))
  ) {
    throw new ASTDecoratorsError(
      `@wrap can only be applied to class methods or properties with ` +
        'functions assigned',
    );
  }
};
