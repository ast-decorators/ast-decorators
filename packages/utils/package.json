{
  "name": "@ast-decorators/utils",
  "version": "0.1.0",
  "description": "Common utils for AST decorators",
  "types": "lib/index.d.ts",
  "scripts": {
    "build": "npm run build:remove && run-p build:js build:dts",
    "build:js": "babel src -d lib -x .ts --ignore src/**/*.d.ts --copy-files --config-file ../../babel.config.json",
    "build:dts": "tsc -p tsconfig.build.json --emitDeclarationOnly",
    "build:remove": "rimraf lib",
    "typecheck": "tsc -p tsconfig.json --noEmit"
  },
  "dependencies": {
    "@ast-decorators/typings": "^0.1.0",
    "@babel/types": "^7"
  }
}
