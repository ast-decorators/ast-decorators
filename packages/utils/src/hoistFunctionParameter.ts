import template from '@babel/template';
import type {Scope} from '@babel/traverse';
import {
  ArrowFunctionExpression,
  functionDeclaration,
  FunctionDeclaration,
  FunctionExpression,
  Identifier,
  isArrowFunctionExpression,
  isFunctionExpression,
  MemberExpression,
  VariableDeclaration,
} from '@babel/types';

export type FunctionParameter =
  | FunctionExpression
  | ArrowFunctionExpression
  | Identifier
  | MemberExpression;

const declarator = template(`const VAR = FUNCTION`);

const hoistFunctionParameter = (
  fn: FunctionParameter,
  name: string,
  scope: Scope,
): readonly [
  Identifier | MemberExpression,
  FunctionDeclaration | VariableDeclaration | undefined,
] => {
  const isArrow = isArrowFunctionExpression(fn);
  const isRegular = isFunctionExpression(fn);

  const interceptorId =
    isArrow || isRegular
      ? scope.generateUidIdentifier(name)
      : (fn as Identifier | MemberExpression);

  return [
    interceptorId,
    isArrow
      ? (declarator({
          FUNCTION: fn,
          VAR: interceptorId as Identifier,
        }) as VariableDeclaration)
      : isRegular
      ? functionDeclaration(
          interceptorId as Identifier,
          (fn as FunctionExpression).params,
          (fn as FunctionExpression).body,
        )
      : undefined,
  ];
};

export default hoistFunctionParameter;
