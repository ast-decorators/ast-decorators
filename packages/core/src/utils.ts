import {ASTDecoratorTransformerOptions} from '@ast-decorators/typings';
import {BabelFileResult, NodePath} from '@babel/core';
import {Binding} from '@babel/traverse';
import {
  CallExpression,
  Decorator,
  Identifier,
  ImportDeclaration,
  ImportDefaultSpecifier,
  ImportNamespaceSpecifier,
  ImportSpecifier,
  isCallExpression,
  isMemberExpression,
  MemberExpression,
  StringLiteral,
} from '@babel/types';

export type ASTDecoratorExclusionOptions = {
  paths?: string[];
  names?: RegExp | string[];
};

export type ASTDecoratorCoreOptions = Readonly<{
  exclude?: ASTDecoratorExclusionOptions;
  transformers?: ASTDecoratorTransformerOptions;
}>;

export type PluginPass<T> = Readonly<{
  cwd: string;
  file: BabelFileResult;
  filename?: string;
  key: string;
  opts?: T;
}>;

const $args = Symbol('args');
const $binding = Symbol('binding');
const $decorator = Symbol('decorators');
const $id = Symbol('id');
const $isCall = Symbol('isCall');
const $object = Symbol('object');

const $handleDecoratorWithArgs = Symbol('handleDecoratorWithArgs');
const $handleDecoratorWithoutArgs = Symbol('handleDecoratorWithoutArgs');
const $handleMemberDecorator = Symbol('handleMemberDecorator');

export default class DecoratorMetadata {
  private [$args]?: readonly NodePath[];
  private [$binding]?: Binding;
  private [$decorator]: NodePath<Decorator>;
  private [$id]: NodePath<Identifier>;
  private [$isCall]: boolean;
  private [$object]?: NodePath<Identifier>;

  public constructor(decorator: NodePath<Decorator>) {
    this[$decorator] = decorator;
    const expression = decorator.get('expression');

    this[$isCall] = isCallExpression(expression);

    if (this[$isCall]) {
      this[$handleDecoratorWithArgs](expression as NodePath<CallExpression>);
    } else {
      this[$handleDecoratorWithoutArgs](
        expression as NodePath<MemberExpression | Identifier>,
      );
    }

    if (this.isFree) {
      this[$binding] = decorator.scope.getBinding(this[$id]!.node.name);
    } else {
      this[$binding] = decorator.scope.getBinding(this[$object]!.node.name);
    }
  }

  public get args(): readonly NodePath[] {
    return this[$args] ?? [];
  }

  public get binding(): Binding | undefined {
    return this[$binding];
  }

  public get bindingId(): NodePath<Identifier> {
    return this.isFree ? this[$id] : this[$object]!;
  }

  public get importSpecifier():
    | NodePath<
        ImportSpecifier | ImportNamespaceSpecifier | ImportDefaultSpecifier
      >
    | undefined {
    return this[$binding] ? (this[$binding]!.path as any) : undefined;
  }

  public get importSource(): NodePath<StringLiteral> | undefined {
    const {importSpecifier} = this;

    if (!importSpecifier) {
      return undefined;
    }

    const declaration = importSpecifier.parentPath as NodePath<
      ImportDeclaration
    >;

    return declaration.get('source');
  }

  public get isCall(): boolean {
    return this[$isCall];
  }

  /**
   * Is decorator identifier a part of MemberExpression?
   */
  public get isFree(): boolean {
    return !this[$object];
  }

  public get property(): string | undefined {
    return this.isFree ? undefined : this[$id].node.name;
  }

  public removeDecorator(): void {
    this[$decorator].remove();
  }

  public removeBinding(): void {
    if (!this[$binding]) {
      return;
    }

    this[$binding]!.referencePaths = this[$binding]!.referencePaths.filter(
      p => p !== this.bindingId,
    );

    if (this[$binding]!.referencePaths.length === 0) {
      const declaration = this.importSpecifier!.parentPath as NodePath<
        ImportDeclaration
      >;

      this.importSpecifier!.remove();

      if (declaration.node.specifiers.length === 0) {
        declaration.remove();
      }
    }
  }

  private [$handleDecoratorWithArgs](
    expression: NodePath<CallExpression>,
  ): void {
    this[$args] = expression.get('arguments');

    const callee = expression.get('callee');

    if (isMemberExpression(callee)) {
      this[$handleMemberDecorator](callee as NodePath<MemberExpression>);
    } else {
      this[$id] = callee as NodePath<Identifier>;
    }
  }

  private [$handleDecoratorWithoutArgs](
    expression: NodePath<MemberExpression | Identifier>,
  ): void {
    if (isMemberExpression(expression)) {
      this[$handleMemberDecorator](expression as NodePath<MemberExpression>);
    } else {
      this[$id] = expression as NodePath<Identifier>;
    }
  }

  private [$handleMemberDecorator](
    expression: NodePath<MemberExpression>,
  ): void {
    this[$object] = expression.get('object') as NodePath<Identifier>;
    this[$id] = expression.get('property') as NodePath<Identifier>;
  }
}
