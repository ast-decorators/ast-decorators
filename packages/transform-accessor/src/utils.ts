import ASTDecoratorsError from '@ast-decorators/utils/lib/ASTDecoratorsError';
import type {SuitabilityFactors} from '@ast-decorators/utils/lib/checkSuitability';
import type {
  ClassMemberMethod,
  ClassMemberProperty,
  PrivacyType,
} from '@ast-decorators/utils/lib/common';
import createPropertyByPrivacy from '@ast-decorators/utils/lib/createPropertyByPrivacy';
import getMemberName from '@ast-decorators/utils/lib/getMemberName';
import {ClassMember} from '@ast-decorators/utils/src/common';
import template from '@babel/template';
import type {NodePath, Scope} from '@babel/traverse';
import {
  ArrayPattern,
  ArrowFunctionExpression,
  Class,
  cloneNode,
  Decorator,
  FunctionDeclaration,
  functionDeclaration,
  FunctionExpression,
  Identifier,
  isArrayPattern,
  isArrowFunctionExpression,
  isClassMethod,
  isClassPrivateMethod,
  isClassPrivateProperty,
  isClassProperty,
  isFunctionExpression,
  isIdentifier,
  isMemberExpression,
  isObjectPattern,
  MemberExpression,
  ObjectPattern,
  PrivateName,
  ThisExpression,
  thisExpression,
  variableDeclaration,
  VariableDeclaration,
  variableDeclarator,
} from '@babel/types';

export type TransformedNode = readonly [
  ClassMemberMethod,
  ReadonlyArray<FunctionDeclaration | VariableDeclaration>,
];

export type TransformAccessorOptions = Readonly<{
  privacy?: PrivacyType;
  singleAccessorDecorators?: SuitabilityFactors;
  transformerPath?: string;
  useClassNameForStatic?: boolean;
}>;

export type AccessorInterceptor = (value: any) => any;
export type AccessorInterceptorNode =
  | FunctionExpression
  | ArrowFunctionExpression
  | Identifier
  | MemberExpression;

export type AccessorMethodCreatorOptions = Readonly<{
  useClassName: boolean;
  preservingDecorators: Decorator[] | null;
}>;

export type AccessorMethodCreator = (
  klass: NodePath<Class>,
  member: NodePath<ClassMember>,
  interceptor: NodePath<AccessorInterceptorNode> | undefined,
  storage: Identifier | PrivateName | undefined,
  options: AccessorMethodCreatorOptions,
) => TransformedNode | null;

export const isGetter = (member: ClassMemberMethod): boolean =>
  member.kind === 'get';
export const isSetter = (member: ClassMemberMethod): boolean =>
  member.kind === 'set';

export const assert = (
  decorator: string,
  member: ClassMember,
  interceptors: ReadonlyArray<AccessorInterceptorNode | undefined>,
): void => {
  if (
    !(isClassProperty(member) || isClassPrivateProperty(member)) &&
    !(
      (isClassMethod(member) || isClassPrivateMethod(member)) &&
      (isGetter(member) || isSetter(member))
    )
  ) {
    throw new ASTDecoratorsError(
      `Applying @${decorator} decorator to something other than property or accessor is not allowed`,
    );
  }

  for (const interceptor of interceptors) {
    if (
      interceptor &&
      !isFunctionExpression(interceptor) &&
      !isArrowFunctionExpression(interceptor) &&
      !isIdentifier(interceptor) &&
      !isMemberExpression(interceptor)
    ) {
      throw new ASTDecoratorsError(
        'Accessor interceptor can only be function, free variable or object property',
      );
    }
  }
};

export const createStorage = (
  klass: NodePath<Class>,
  member: ClassMemberProperty,
  privacy: PrivacyType = 'hard',
): ClassMemberProperty =>
  createPropertyByPrivacy(privacy, getMemberName(member), klass, {
    // @ts-expect-error: "static" is not listed in d.ts
    static: member.static,
    value: member.value,
  });

const declarator = template(`const VAR = FUNCTION`);

export const ownerNode = (
  klass: Class,
  useClassName: boolean,
): Identifier | ThisExpression =>
  useClassName && klass.id ? cloneNode(klass.id) : thisExpression();

export const prepareInterceptor = (
  klass: NodePath<Class>,
  interceptor: AccessorInterceptorNode,
  type: 'get' | 'set',
): readonly [
  Identifier | MemberExpression,
  FunctionDeclaration | VariableDeclaration | null,
] => {
  const isArrow = isArrowFunctionExpression(interceptor);
  const isRegular = isFunctionExpression(interceptor);

  const interceptorId =
    isArrow || isRegular
      ? klass.parentPath.scope.generateUidIdentifier(type)
      : (interceptor as Identifier | MemberExpression);

  return [
    interceptorId,
    isArrow
      ? (declarator({
          FUNCTION: interceptor,
          VAR: interceptorId as Identifier,
        }) as VariableDeclaration)
      : isRegular
      ? functionDeclaration(
          interceptorId as Identifier,
          (interceptor as FunctionExpression).params,
          (interceptor as FunctionExpression).body,
        )
      : null,
  ];
};

export const unifyValueParameter = (
  scope: Scope,
  rawValue: Identifier | ArrayPattern | ObjectPattern,
): [Identifier, VariableDeclaration?] => {
  if (isObjectPattern(rawValue) || isArrayPattern(rawValue)) {
    const value = scope.generateUidIdentifier('value');

    return [
      value,
      variableDeclaration('const', [variableDeclarator(rawValue, value)]),
    ];
  }

  return [rawValue];
};
