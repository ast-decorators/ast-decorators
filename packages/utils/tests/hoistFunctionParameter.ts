import {NodePath, traverse} from '@babel/core';
import generate from '@babel/generator';
import {
  CallExpression,
  Class,
  Decorator,
  Identifier,
  isClass,
  isIdentifier,
  isMemberExpression,
} from '@babel/types';
import {parseToAST as _parseToAST} from '../../../utils/testing';
import hoistFunctionParameter, {
  FunctionParameter,
} from '../src/hoistFunctionParameter';
import options from './fixtures/getMemberName/options';

const parseToAST = async (fixture: string): ReturnType<typeof _parseToAST> =>
  _parseToAST(__dirname, 'hoistFunctionParameter', fixture, options);

const run = async (
  fixture: string,
  callback: (fn: FunctionParameter, classPath: NodePath<Class>) => void,
): Promise<void> => {
  const ast = await parseToAST(fixture);

  await new Promise((resolve) => {
    traverse(ast!, {
      Decorator(path: NodePath<Decorator>) {
        const [fn] = (path.node.expression as CallExpression).arguments;

        // eslint-disable-next-line callback-return
        callback(
          fn as FunctionParameter,
          path.findParent(({node}) => isClass(node)) as NodePath<Class>,
        );
        resolve();
      },
    });
  });
};

describe('@ast-decorators/utils', () => {
  describe('hoistFunctionParameter', () => {
    it('hoists the arrow function parameter', async () => {
      await run('parameter-arrow', (fn, classPath) => {
        const [id, declaration] = hoistFunctionParameter(
          fn,
          'param',
          classPath.parentPath.scope,
        );

        expect(isIdentifier(id)).toBeTruthy();
        expect((id as Identifier).name).toBe('_param');
        expect(declaration).not.toBeUndefined();
        expect(generate(declaration!).code).toMatchSnapshot();
      });
    });

    it('hoists the regular function parameter', async () => {
      await run('parameter-regular', (fn, classPath) => {
        const [id, declaration] = hoistFunctionParameter(
          fn,
          'param',
          classPath.parentPath.scope,
        );

        expect(isIdentifier(id)).toBeTruthy();
        expect((id as Identifier).name).toBe('_param');
        expect(declaration).not.toBeUndefined();
        expect(generate(declaration!).code).toMatchSnapshot();
      });
    });

    it('gets the identifier parameter id', async () => {
      await run('parameter-identifier', (fn, classPath) => {
        const [id, declaration] = hoistFunctionParameter(
          fn,
          'param',
          classPath.parentPath.scope,
        );

        expect(isIdentifier(id)).toBeTruthy();
        expect((id as Identifier).name).toBe('fn');
        expect(declaration).toBeUndefined();
      });
    });

    it('gets the MemberExpression parameter id', async () => {
      await run('parameter-member-expression', (fn, classPath) => {
        const [id, declaration] = hoistFunctionParameter(
          fn,
          'param',
          classPath.parentPath.scope,
        );

        expect(isMemberExpression(id)).toBeTruthy();
        expect(generate(id).code).toBe('utils.fn');
        expect(declaration).toBeUndefined();
      });
    });

    it('changes the id of the hoisted parameter to be unique', async () => {
      await run('parameter-identifier-unique-name', (fn, classPath) => {
        const [id] = hoistFunctionParameter(
          fn,
          'someUniqueVar',
          classPath.parentPath.scope,
        );

        expect(isIdentifier(id)).toBeTruthy();
        expect((id as Identifier).name).toBe('_someUniqueVar2');
      });
    });
  });
});
