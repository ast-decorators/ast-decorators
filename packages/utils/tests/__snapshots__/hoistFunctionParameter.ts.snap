// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@ast-decorators/utils hoistFunctionParameter hoists the arrow function parameter 1`] = `"const _param = (baz: number) => baz * 10;"`;

exports[`@ast-decorators/utils hoistFunctionParameter hoists the regular function parameter 1`] = `
"function _param(baz: number) {
  return baz * 10;
}"
`;
